### ISSUE TITLE
Add support for CycloneDX Output format

### ISSUE THREAD

**Issue by eastmadc** [nexB/scancode-toolkit#1888]
Similar to the support of exporting to SPDX output format.  It would be helpful to add support to export into the increasingly popular CycloneDX format.

[Specification](https://github.com/CycloneDX/specification)
[Example Project](https://github.com/CycloneDX/cyclonedx-node-module)

**Comment by agschrei** [comment]
Having support for this output format would be a great help to anybody that wants to ingest ScanCode results into downstream tools like DependencyTrack or enrich them with data from other scanners/sources.  

I'm assuming many people have a business need for this, so I'm leaving this comment here in the hopes that it gains traction.  

Also, looking at the plugin architecture, it seems to me like implementing this as a post-scan plugin is possible.   
Could anybody with more hands-on experience on ScanCode plugins comment on that perhaps?

**Comment by pombredanne** [comment]
@eastmadc @agschrei yes that would be fairly straightforward to implement especially since CycloneDX embases Package URL that were born here. The approach is to implement an output plugin like in https://github.com/nexB/scancode-toolkit/blob/develop/src/formattedcode/output_yaml.py
You can see the simple commit stream here https://github.com/nexB/scancode-toolkit/issues/2208 covering the whole thing (or this simple single commit https://github.com/nexB/scancode-toolkit/commit/310f4501a720fdb201e86e4d63eafb3e22e347b3 )

And this library is likely the way to go too https://github.com/CycloneDX/cyclonedx-python

**Comment by pombredanne** [comment]
@stevespringett FYI ^ :)

BTW in  https://github.com/CycloneDX/cyclonedx-python/blob/master/setup.py there are several deps which are pinned which makes it somewhat impractical to reuse this as a library. I will likely provide a patch for your consideration to makes this easier to reuse.

**Comment by pombredanne** [comment]
@agschrei Are you willing to chip in to help implement this? I can provide hand holding!

**Comment by agschrei** [comment]
@pombredanne I'll hopefully find some time this weekend to look into it. I can't guarantee anything just yet because of university exams, but will report back on Monday.   

Given you know the codebase like the back of your hand, how do you estimate this feature in terms of effort/time required?

**Comment by pombredanne** [comment]
> I'll hopefully find some time this weekend to look into it. I can't guarantee anything just yet because of university exams, but will report back on Monday.

That's very gentle of you!

> Given you know the codebase like the back of your hand, how do you estimate this feature in terms of effort/time required?

The easy part is likely to be the writing of the Cyclone file. The harder may be likely how to align the data with the Cyclone data model: for instance what about file-level details vs. packages? ScanCode collects also a lot of details that would likely not map with Cyclone at all.

Now, say as a first pass that this covers only detected packages (with the `--package` option) I would estimate this to be about 1 to 2 days work with at least one simple test.

**Comment by coderpatros** [comment]
@agschrei if you want any advice about aligning the data to the CycloneDX spec feel free to hit me up on the [CycloneDX Slack](https://cyclonedx.org/slack/invite)

@pombredanne I'm happy to provide feedback on the implementation too.

**Comment by coderpatros** [comment]
I've just gone back over this thread and the issue @pombredanne raised regarding pinned dependency versions in the CycloneDX Python implementation https://github.com/CycloneDX/cyclonedx-python/issues/211

The Python implementation only supports a subset of the spec. So trying to use it for this might be a hindrance. We of course welcome PRs. Just don't want you to be held up waiting for PRs to be merged when it's so easy to just implement it in Python. Well, JSON format is trivial to implement in Python at least. I haven't implemented XML or Protocol Buffers in Python myself. They should be pretty easy too. But JSON is _really_ easy.

**Comment by agschrei** [comment]
@coderpatros thanks a ton for the offer, it's helpful to have a point of contact in case I run into trouble.

Concerning the lib, I also noticed that it doesn't provide any support for CycloneDX 1.3 yet, so perhaps also not a good fit in that respect. 

I'll have to look into [cyclonedx-python](https://github.com/CycloneDX/cyclonedx-python) to see how much additional work support for revision 1.3 (or rather the parts of it we want/need) requires. Personally, I'd be a lot happier with improving the library rather than rolling our own implementation for the ScanCode plugin, but it's of course also a matter of time and resources. 

@pombredanne  
what's your preferred flavor of CycloneDX to support here?  

I could also introduce two different flags like `--cyclonedx` for XML and `--cyclonedx-json` for JSON, similar to what you are offering in the SPDX output plugin.

**Comment by pombredanne** [comment]
>  what's your preferred flavor of CycloneDX to support here?

Let's go with JSON for a start. I like its rounded curlies more than XML's pointies. 

> I could also introduce two different flags like --cyclonedx for XML and --cyclonedx-json for JSON, similar to what you are offering in the SPDX output plugin.

This works fine.


> Personally, I'd be a lot happier with improving the library rather than rolling our own implementation for the ScanCode plugin, but it's of course also a matter of time and resources.

Same. I suggest though to start with something concrete here first and this can be extracted and contributed upstream afterwards.

@coderpatros thanks for chiming in !

**Comment by coderpatros** [comment]
> Let's go with JSON for a start. I like its rounded curlies more than XML's pointies.

As good a reason as any XD

**Comment by agschrei** [comment]
> Let's go with JSON for a start. I like its rounded curlies more than XML's pointies.  

I can get behind that line of reasoning.

> Same. I suggest though to start with something concrete here first and this can be extracted and contributed upstream afterwards.

Sounds good to me, I've already gotten started on some boilerplate, next step is doing the actual mapping to CycloneDX.

**Comment by pombredanne** [comment]
@agschrei gentle ping :) 'sup on your side? Any help needed?

**Comment by agschrei** [comment]
@pombredanne apologies for falling silent here,  
so far I haven't been able to dedicate as much time to this as I had originally hoped, but I've at least had time to familiarize myself with the other output plugins, contribution guidelines and to set up some scaffolding for the CycloneDX plugin.

I'd assume that the majority of details can best be nailed down once I have a working prototype, which I'm going to spend some time on today.  

That having been said there is one question that came up browsing the other output plugins:  
How would you like entries that have `scan_errors` to be handled?   

The debian and spdx plugins seem to ignore these as far as I can tell.  

If the goal is to produce a strict CycloneDX BOM there might not be a good way to reflect them in the BOM anyway, but intuitively I would at least like to print a warning to the logs that some components had errors.     

Another alternative would be using the non-strict schema which allows definition of custom data attributes,  but I think that would only be of limited use.  

What's your preferred option here?

**Comment by pombredanne** [comment]
@agschrei let's not worry about errors for now.

@coderpatros I see you contributed CycloneDX support to Tern in https://github.com/tern-tools/tern/commit/9bbb5dd3100a37956511cbe89af960447b578f24 ... is there something that could be extracted from there as a base?

**Comment by coderpatros** [comment]
@agschrei if there is information ScanCode captures, that doesn't quite fit, you can use custom properties.

https://cyclonedx.org/use-cases/#properties--name-value-store

@pombredanne most of the Tern changes were to work in with their data model and way of doing things. I also took a minimal approach. So it can probably be used as a guide for the implementation here. But probably not very useful to lift straight out.

@agschrei if you can get the scaffolding in place and a working prototype, I'm happy to have a look at it and help with mapping ScanCode data to the CycloneDX spec.

**Comment by agschrei** [comment]
@pombredanne
while I'm not done yet, there are a few questions that came up while working on this, I was hoping you could enlighten me on some of them. 

Disclaimer: So far no code-cleanup has happened, I want to get to that once we have a result that everybody is happy with content-wise.

1. I know the pythonic way for doing this would probably just be to shove everything into a nested dictionary and be done with it, but I've tried to extract CycloneDx structs into class definitions similarly to what the [schema](https://json-schema.app/view/%23?url=https%3A%2F%2Fraw.githubusercontent.com%2FCycloneDX%2Fspecification%2Fmaster%2Fschema%2Fbom-1.3.schema.json) defines wherever possible.  
Given the plan (already discussed above) to extract the CycloneDx portion into a standalone package, is this an approach you are okay with?

2. Is it safe to assume that [packagedcode/models.py](https://github.com/nexB/scancode-toolkit/blob/develop/src/packagedcode/models.py) covers all properties a package can have or is there some additional extension happening somewhere else in packagedcode?

3. The tools section of the BOM metadata currently sets "nexB Inc." as vendor, is this okay for you?

One additional thing I want to tackle next is mapping copyright information (author, maintainer, etc.) to a CycloneDx Bom.

Since GitHub doesn't allow me to attach the json files directly, I had to zip them, but 
[this-archive](https://github.com/nexB/scancode-toolkit/files/7093972/results.zip) contains both a json-pp.json and cyclonedx.json report for the same scan (of a small npm project).  
The cyclonedx file has been prettified for easier side-by-side comparison, the output plugin currently builds a compact file.

@coderpatros I would really appreciate your insight on the current quality of the mapping here if you could spare a few minutes.

My guess is the license [expression](https://json-schema.app/view/%23/%23%2Fdefinitions%2Fcomponent/%23%2Fdefinitions%2FlicenseChoice?url=https%3A%2F%2Fraw.githubusercontent.com%2FCycloneDX%2Fspecification%2Fmaster%2Fschema%2Fbom-1.3.schema.json) mapping can still be improved (I currently just carry over the `license_expression` key)

**Comment by coderpatros** [comment]
> Given the plan (already discussed above) to extract the CycloneDx portion into a standalone package

I wouldn't worry about this. Turns out Sonatype have been working on a Python library they are going to donate to the CycloneDX community.

**Comment by coderpatros** [comment]
The output looks pretty good so far.

You could use component scope to indicate packages that are/are not included in released artifacts.

What's the difference between `license_expression` and `declared_license` here?

**Comment by agschrei** [comment]
> I wouldn't worry about this. Turns out Sonatype have been working on a Python library they are going to donate to the CycloneDX community.  

That's awesome news. I'm assuming [this](https://github.com/sonatype-nexus-community/cyclonedx-python-lib)  is the project?

> You could use component scope to indicate packages that are/are not included in released artifacts.   

That's still on my list of TODOs.  

Ideally I want to map the requirements that ScanCode outputs per package to the `Dependency` type that CycloneDx exposes, but to do that I will need to:  
a) map the requirements to a concrete package (they are usually given in terms of compatible versions not resolved to a specific package)  
b) assign bom-ref keys to all components  

@coderpatros  the only constraint the  json schema for CycloneDx 1.3  imposes on bom-ref is that it needs to be a string. Is there some best practice I should follow here, or could I feasibly also use a package-url as a bom-ref?   

> What's the difference between `license_expression` and `declared_license` here?   

In this particular case the declared package license and the `license_expression` are the same for all npm packages that were discovered, but if I'm not mistaken, for some other package types, the `declared_license` field might also hold a list of licenses or a JSON object with additional license info.  
I will add some logic to filter out duplicate entries so we don't cover the same license multiple times.

Apart from re-ordering a few things here and there I mainly see the following open tasks:  

- [x] map copyright info and the author field
- [x] express the dependency tree (including scope) by mapping requirements to CycloneDx `Dependency` entries
- [x] improve the license mapping
- [x] write an XML serializer for the defined types

**Comment by agschrei** [comment]
The output plugin now maps dependencies, it also provides working XML output.
[results.zip](https://github.com/nexB/scancode-toolkit/files/7117788/results.zip)

For simplicity's sake package-urls are used as bom-refs, I assume that's okay given that the [examples](https://cyclonedx.org/use-cases/#dependency-graph) do the same thing

@coderpatros is there anything I am still missing here?

**Comment by stevespringett** [comment]
> For simplicity's sake package-urls are used as bom-refs, I assume that's okay given that the examples do the same thing

Absolutely. Package URLs are perfectly fine. However, the bom-ref is required to be unique within the context of the BOM. Under normal circumstances, this isn't a problem. Just something to be aware of.

**Comment by coderpatros** [comment]
That is the correct project. First official CycloneDX release is up https://pypi.org/project/cyclonedx-python-lib/

**Comment by tim-jagenberg-piscada** [comment]
I am trying to use the CycloneDX format from the current `develop` branch, but I'm getting the following error when it tries to write out the collected results:

`./scancode --license --package --cyclonedx test.json PATH_TO_THE_TARGET`

```
ERROR: failed to run output plugin: cyclonedx:
Traceback (most recent call last):
  File "/home/tim/git/scancode-toolkit/src/scancode/cli.py", line 1057, in run_codebase_plugins
    plugin.process_codebase(codebase, **kwargs)
  File "/home/tim/git/scancode-toolkit/src/formattedcode/output_cyclonedx.py", line 735, in process_codebase
    bom = CycloneDxBom.from_codebase(codebase)
  File "/home/tim/git/scancode-toolkit/src/formattedcode/output_cyclonedx.py", line 634, in from_codebase
    components = list(CycloneDxComponent.from_packages(packages))
  File "/home/tim/git/scancode-toolkit/src/formattedcode/output_cyclonedx.py", line 304, in from_packages
    comp = CycloneDxComponent.from_package(package)
  File "/home/tim/git/scancode-toolkit/src/formattedcode/output_cyclonedx.py", line 271, in from_package
    value=f'WARNING: component skipped in CycloneDX output: {self!r}'
NameError: name 'self' is not defined
```

**Comment by pombredanne** [comment]
@tim-jagenberg-piscada Thank you ++ for the report!

**Comment by eastmadc** [comment]
Implemented by #2698

**Comment by eastmadc** [comment]
Implemented by #2698
